forem/forem
# ✂️  Remove reach combobox ✂️

## What type of PR is this? (check all applicable)


## Description

This PR removes the frontend dependency `@reach/combobox`, which we used purely for the mention autocomplete feature. 

### Motivation

I wanted to remove this package because

1. Following work on the tag autocomplete feature I felt like we had a better sense of what would need to be done to achieve an accessible custom implementation
2. We had to do a lot of hacks around the library to make the feature work before (it is designed for small searchable inputs, not large text areas that _most of the time_ are not suggesting anything)
3. The hacks we did meant that it was difficult to get your head around, difficult to update or extend, and our tests were very tightly coupled to the implementation under the hood
4. Fewer dependencies === smaller bundle size 😄 

### Impact

In almost every way, the mention autocomplete feature in the editor and comment areas should function the same as before. The real difference is that whereas before we switched between two text areas during usage (one when in "suggesting" mode, and one for the rest of the time) , we now always stay in the one text area but dynamically enhance it with the proper ARIA for the "suggesting" mode when triggered by typing `@`.

I believe this to be better for a11y, more robust, and also just way more testable 😅 

I've also taken steps to generalise the component a bit more so that we can more easily extend it in future. For example, the character which starts the search behaviour can now be set in props, rather than hard-coded to `@`, and a Preact component can be passed in to determine the layout and appearance of fetched "suggestions" 

## Related Tickets & Documents

N/A

## QA Instructions, Screenshots, Recordings

This is one of those UI-heavy changes where we really ought to test in multiple browsers, and on mobile devices as well as desktop. I've personally tested in:

- Firefox / Chrome / Safari
- Chrome on Android mobile
- VoiceOver with Safari for screen reader checks

Test steps:

- Go to create a new post, and try `@mentioning` a user. It would be a good idea to make sure your database has multiple users with similar names, so that you can check quite a few appearing in a list.
- Try the same in a post's comment text area
- Try the same in a _reply_ to an existing comment
- Experiment with positioning of the `@mention` (end or middle of the text/line) - the dropdown should appear under the mention in all cases
- Make sure you can still interact with the text areas in all the normal ways (e.g. in the editor try drag/dropping an image)

### UI accessibility concerns?

Yes! As a brief summary:

- When a user types `@` an `aria-live` announces "type to search for a user", and the text area is enhanced with the appropriate ARIA and keyboard behaviours for the [combobox pattern](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/)
- When the suggestions list is visible, you should be able to use the Up/Down arrow keys + Enter to select an option (in which case it should behave the same as if you clicked the option)
- When the suggestions list is visible, you should be able to press `Escape` to hide it
- During normal usage (no `@` sign at the beginning of the current word), the text area should just be a normal, plain ol' text area. There should be no "combobox" ARIA 


## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [X] This change does not need to be communicated, and this is why not: no change to functionality


## [optional] What gif best describes this PR or how it makes you feel?

![This went well, I think](https://media.giphy.com/media/kDWFQj74Ka4vQiUhw5/giphy-downsized.gif)
